### PR TITLE
channelIfGrace

### LINKED ISSUE #102: Grace periods and validUntil
disallow updating the state if we're past `validUntil`, or `validUntil-channel.spec.gracePeriod` if `gracePeriod` is set

in other words, we should not allow balances to be updated if the channel is expired or in a grace period; the grace period is needed to allow some time for publishers to withdraw their final balances from the channel, before the channel is expired on-chain

in practice, we should implement this by disallowing all events other than channel close (#48)

### PR DESCRIPTION
**Pull request by samparsky**
Fixes #102 #105

### CODE REVIEW

File: middlewares/channel.js [hunk 1]
```diff
@@ -40,4 +40,16 @@ function channelIfActive(req, res, next) {
 	channelIfFind({ _id: req.params.id, 'spec.validators.id': req.whoami }, req, res, next)
 }
 
-module.exports = { channelLoad, channelIfExists, channelIfActive }
+// requires channelLoad
+function channelIfGrace(req, res, next) {
+	const { channel } = req
+	const currentTime = Date.now()
+	if (channel.spec.gracePeriod || currentTime > channel.validUntil) {
```
Comment (Ivshti) [inline middlewares/channel.js]: this seems wrong - the value of `channel.spec.gracePeriod > channel.validUntil` will always be constant

should be `currentTime - gracePeriod`

File: middlewares/channel.js [hunk 2]
```diff
@@ -40,4 +40,23 @@ function channelIfActive(req, res, next) {
 	channelIfFind({ _id: req.params.id, 'spec.validators.id': req.whoami }, req, res, next)
 }
 
-module.exports = { channelLoad, channelIfExists, channelIfActive }
+// requires channelLoad
+function channelIfGrace(req, res, next) {
```
Comment (Ivshti) [inline middlewares/channel.js]: rename to `withdraw` period

File: routes/schemas.js [hunk 3]
```diff
@@ -47,9 +47,55 @@ module.exports = {
 		// we want it to be stored in MongoDB as a number
 		validUntil: Joi.number().required(),
 		creator: creator(cfg),
+		created: Joi.number(),
+		ipfs: Joi.string().regex(/^ipfs:/),
```
Comment (Ivshti) [inline routes/schemas.js]: these properties should definitely not be here; anything new should be in the spec

File: routes/schemas.js [hunk 4]
```diff
@@ -47,9 +47,55 @@ module.exports = {
...
+		title: Joi.string(),
+		description: Joi.string(),
+		archived: Joi.boolean(),
+		modified: Joi.number(),
 		spec: Joi.object({
```
Comment (Ivshti) [inline routes/schemas.js]: let's change this to enforce the rules for `validators` and `withdrawPeriodStart` but allow *any other properties*; and let's move the adunits schema into the market

File: middlewares/channel.js (hunk 2 above)
Comment (samparsky) [reply middlewares/channel.js]: done

File: routes/schemas.js (hunk 4 above)
Comment (samparsky) [reply routes/schemas.js]: done

**Pull request merged**